### ISSUE TITLE
unable to install c++ plugin

### ISSUE THREAD

**Issue by quantrpeter** [apache/netbeans#6979]
### Apache NetBeans version

Apache NetBeans 20

### What happened

<img width="1100" alt="image" src="https://github.com/apache/netbeans/assets/50471577/1fefe9d0-e39e-439c-8820-b3ee4e51c64a">

### How to reproduce

install c++ plugin

### Did this work correctly in an earlier version?

No / Don't know

### Operating System

Mac

### JDK

21

### Apache NetBeans packaging

Apache NetBeans provided installer

### Anything else

nothing

### Are you willing to submit a pull request?

No

**Comment by bitrunner** [comment]
Like the dialog says, unpack200 is required to extract old netbeans modules like CND.  Java 14 removed unpack200, so one way to resolve this is to install JDK 8 or 11 and then use the bottom left button to "Choose Unpack200..." on the dialog to select the unpack200 executable to complete the CND install.

**Comment by quantrpeter** [comment]
thanks @bitrunner  May I know where is the c++ plugin? may be i can help to revamp it

**Comment by bitrunner** [comment]
Mostly @vieiro (with some clueless flailing around help from me) has been working to bring CND back to life.  You can find the current state on the [cnd branch](https://github.com/apache/netbeans/tree/cnd) of the main netbeans repository here on github.

The module can be found in the cnd directory.  There is some conflict/overlap with the newer cpplite module that has to be resolved before CND can be incorporated back into netbeans proper.  There have been some discussions regarding that on the [mailing list](https://lists.apache.org/list.html?dev@netbeans.apache.org).   All help is appreciated.

**Comment by matthiasblaesing** [comment]
I agree with @mbien we can't stay in the past. If there is really someone caring, he/she can take the CND plugins, unpack200 them and create a new distribution.

**Comment by vieiro** [comment]
> Mostly @vieiro (with some clueless flailing around help from me) has been working to bring CND back to life. You can find the current state on the [cnd branch](https://github.com/apache/netbeans/tree/cnd) of the main netbeans repository here on github.
> 
> The module can be found in the cnd directory. There is some conflict/overlap with the newer cpplite module that has to be resolved before CND can be incorporated back into netbeans proper. There have been some discussions regarding that on the [mailing list](https://lists.apache.org/list.html?dev@netbeans.apache.org). All help is appreciated.

Well, to be fair, many people have been pushing the cnd branch, including @matthiasblaesing  and @jlahoda, among others. There're incompatibilities with cpplite, but the cnd branch can be built (using `ant -Dcluster.config=cnd`). I don't understand how cpplite is being used, so I really can't help improving it to make it compatible with cnd.

**Comment by jlahoda** [comment]
FWIW, I've started blending cpplite into cnd some time ago, but not quite done yet, I'm afraid.

**Comment by bitrunner** [comment]
> FWIW, I've started blending cpplite into cnd some time ago, but not quite done yet, I'm afraid.

@jlahoda, I would like to help get this done.

**Comment by jlahoda** [comment]
FWIW, my current sketch to blend cpplite into CND is here:
https://github.com/jlahoda/netbeans/tree/cnd-cpplite-blend

It removes the cpplite.editor (not strictly necessary, but since it is never used anymore, there's no big reason why we should keep it), and rebases cpplite.debugger on top of `cnd.debugger.common2` (i.e. it is another frontend to `cnd.debugger.common2`, just as `cnd.debugger.gdb2` is).

There may be an issue with the debugger - it is used in the VS Code extension, and the CND's debugger seems to require many dependencies. Which would make the extension even bigger than it is now, for a little benefit. One solution would be to try to limit the dependencies of the `cnd.debugger.common2`. Alternative might be to exclude the cpplite.debugger from the standard distro, and keep it only for the VS Code extension and/or other special uses.

Some things that remain, from the top of my head:
- on cpplite side:
 -- verify the native image debugging still works with cpplite.debugger and fix it as needed (unless the changes will be reverted)
 -- decide what to do with the Lightweight C/C++ project - it is sometimes useful as a "freeform" project, and as a base specialized projects, but we probably don't want to push that too much to our users
- on CND side:
 -- support for ccls (this shouldn't really be difficult, and I am not clear why there's a pushback against this)
 -- cnd.lsp must allow other projects to provide the (location of the) compilation database. Besides the Lightweight project, the `java/java.openjdk.project` can provide the (location of the) compilation database.

**Comment by vieiro** [comment]
>     * on CND side:
>       -- support for ccls (this shouldn't really be difficult, and I am not clear why there's a pushback against this)
>       -- cnd.lsp must allow other projects to provide the (location of the) compilation database. Besides the Lightweight project, the `java/java.openjdk.project` can provide the (location of the) compilation database.

`cnd.lsp` should be probably renamed to `cnd.makeproject.lsp`, as it's specific for make based projects. Adding support for `ccls` shouldn't be difficult. Currently a single `clangd` instance can support different projects, I'm not sure if `ccls` is able to that. IIRC ccls didn't support different projects last time I looked at it.

**Comment by bitrunner** [comment]
IMHO, it would be great if cnd.lsp wasn't specific to make based projects.  I would like to use cnd.lsp with the meson project support I added too.  Meson produces its own compilation database that should be fed to the server.  Does it make any sense to separate the makefile specific compilation database generation from perhaps a more general cnd.lsp that would require a compilation database path to be given to it?  I can work on that if there is consensus on that being a reasonable path forward.

Do we need some kind of plan to try to have some kind of CND available in NB 22 or is that just a waste of time and effort should be focused on finishing up Jan's work?

**Comment by vieiro** [comment]
> IMHO, it would be great if cnd.lsp wasn't specific to make based projects. I would like to use cnd.lsp with the meson project support I added too. Meson produces its own compilation database that should be fed to the server. Does it make any sense to separate the makefile specific compilation database generation from perhaps a more general cnd.lsp that would require a compilation database path to be given to it? I can work on that if there is consensus on that being a reasonable path forward.

Yep, that' s a good idea, I think. We could lay out things like so:

- "cnd.makeproject.lsp", responsible for generating the compilation database for NetBeans "makeproject"-type projects.
- "cnd.lsp", responsible for launching one or more LSP servers in background using "ide/lsp.client"
    - Projects should define the "flavor" of the LSP server they wan to run (say "clangd" or "ccls"). 
    - This could be as simple returning an implementation of an interface in the project's Lookup. This interface could return the flavor of the LSP server (ccls or clangd) and maybe the location of the compilation database relative to the project's root, for instance.
    - Some LSP servers can handle multiple projects, others cannot. This may require more investigation. AFAIK a single "clangd" process can handle different projects (cnd.lsp launches a single clangd process), whereas "ccls" requires a process per project. "cnd.lsp" should handle this gracefully.

BTW, "ide/lsp.client" will need some refactoring because of race conditions. The "rust-analyzer" LSP server is complaining that we're sending open document notifications when the server has not been initialized yet.

> 
> Do we need some kind of plan to try to have some kind of CND available in NB 22 or is that just a waste of time and effort should be focused on finishing up Jan's work?

I think we should wait to see if Jan can do some progress in improving cpplite with the cnd gdb debuggers. All help should go on Jan's work, I think.

Meanwhile I'll try to upgrade NB's "antlr4" support, drop the obsolete tomlj and move on the pending Rust PR as time permits (little time for this lately, I'm afraid).

**Comment by obspsr** [comment]
Hi NetBeans people,
I just suggest that you warn our **NetBeans Users** (to avoid time loss), **in the C/C++ site section**, that
- due the deadline of JDK 11 --> Netbeans 21 is the last version usable for C++.
- Added the fact that git-remote (which was essential in the devel process on distant
  32Bits industrial racks) is also working only in 21 version ...
It seems that C/C++ is lost for a long time for NetBeans, very bad news, as NB has reached
a really very high level for PHP devel ... :-(((
Personally, I am moving all my C/C++ projects to KDevelop.
Thanks for coop,
Serge.

**Comment by X547** [comment]
> due the deadline of JDK 11 --> Netbeans 21 is the last version usable for C++.

I tried Netbeans 21 with JDK 11 and there are no option to install C++ support except CCPLite Kit.

**Comment by obspsr** [comment]
You have to go ... Tools -> Plugins and there **Uninstall** the C/C++ lite.
after that, go to Settings and **Add** a new portal as "NetBeans 8.2 Plugin Portal" 
     https://updates.netbeans.org/netbeans/updates/8.2/uc/final/distribution/catalog.xml.gz
and check his mark "Active"...

After that, in Available Plugins you will find the C++ Suite.
**Install** it and during the 2 minutes install you will have to **accept** and show the path of **JDK-11/bin**
(mine is /usr/java/jdk-11/bin/) where lives pack/unpack200 programs.
The installation finished you're done !
And the best, via the Services tab, you can create and work distant C/C++ projects with git-distant ...
Best regards,
Serge.

**Comment by obspsr** [comment]
My full sequence works with Netbeans 21 ... and sure on LINUX (for M$ I don´t care !)
Regards, Serge

**Comment by X547** [comment]
What is the status of https://github.com/apache/netbeans/tree/cnd branch? How to install it? Is it still a plugin or it become a core part of Netbeans? It seems have Meson support that I am interested in.

**Comment by vieiro** [comment]
> What is the status of https://github.com/apache/netbeans/tree/cnd branch? How to install it? Is it still a plugin or it become a core part of Netbeans? It seems have Meson support that I am interested in.

Make sure you have JDK11 (and bigger?) and `ant` (10.>) in your `$PATH` and then

```bash
$ git clone https://github.com/apache/netbeans.git
$ cd netbeans
netbeans$ ant -Dcluster.config=cnd
netbeans$ ant tryme
```

You should install a recent C/C++ LSP server such as `clangd` or `ccls`.

**Comment by bevin-hansson** [comment]
> You should install a recent C/C++ LSP server such as `clangd` or `ccls`.
Is the LSP requirement a later addition? The older C++ plugin doesn't seem to require anything like that, you just point it to the code and it works out of the box.

**Comment by vieiro** [comment]
> > You should install a recent C/C++ LSP server such as `clangd` or `ccls`.
> 
> Is the LSP requirement a later addition? The older C++ plugin doesn't seem to require anything like that, you just point it to the code and it works out of the box.

The "older C++ plugin" cannot be incorporated directly into Apache NetBeans due to license constraints and code provenance issues. The LSP servers are substitute to those parts. And they're easier to maintain and more accurate.

**Comment by X547** [comment]
> The "older C++ plugin" cannot be incorporated directly into Apache NetBeans due to license constraints

Does it mean that license is proprietary/NDA or it is just incompatible with Apache Netbeans? If second, it can be developed in separate repo and packaged as plugin.

**Comment by vieiro** [comment]
> > The "older C++ plugin" cannot be incorporated directly into Apache NetBeans due to license constraints
> 
> Does it mean that license is proprietary/NDA or it is just incompatible with Apache Netbeans? If second, it can be developed in separate repo and packaged as plugin.

To begin with not all modules in the original CND were donated to the ASF, so the code didn't even compile (that'd be the first commit in the cnd branch). Then there where license incompatibilities and obscure code provenance. Then also the difficulty of maintaining such an old/big codebase (which has been evident during the years). 

You can see the whole history in both the "cnd" branch commit history and in the Apache NetBeans mailing list archives.

**Comment by X547** [comment]
> You should install a recent C/C++ LSP server such as clangd or ccls.

It worked for me for default Makefile based projects, but not for Meson projects. Is it not supported yet or I am missing something?

**Comment by matthiasblaesing** [comment]
For meson have a look here: https://github.com/apache/netbeans/pull/6859

**Comment by mrechte** [comment]
Unfortunately starting NB22 with openjdk 11 produces an error:
![Capture d’écran du 2024-10-17 17-10-37](https://github.com/user-attachments/assets/570f05ec-63a0-424f-a4a5-810b424554d5)

**Comment by matthiasblaesing** [comment]
@mrechte yes NB22 requires JDK17 but you can choose the path to unpack200 from JDK 11 to install older plugins.

**Comment by PetrMiksik** [comment]
I found a workaround how to use the CND plugin in NetBeans 21 and newer. It's not exactly a clean procedure, but it works for me (even in NetBeans 25).
Everytime I upgraded NetBeans, I used the option to transfer the configuration from the previous version. In NetBeans 21, activation of the CND plugin failed. So I installed a clean version and instead of transferring the configuration, I simply copied whole configuration folder from the previous version. I got an error that CND module failed to activate, and even in the list of installed plugins looks like deactivated, but it's working.
I understand that such a procedure may have undesirable side effects, so later I copied only the CND module files with their dependencies and it works as well.

**Comment by vinothv84** [comment]
@PetrMiksik I am a long term user of CND plugin. I am also using NetBeans 21. If possible, Can you please elaborate on what files/folder you copied exactly to make it work on latest NetBeans? BTW, I am using NetBeans on a windows PC.

**Comment by PetrMiksik** [comment]
@vinothv84 The following modules are required: **clank, cnd**, and **dlight**
NetBeans configuration folder is usually located here (on Windows):
C:/Users/_»LOGGED-USER«_/AppData/Roaming/NetBeans/_»NB-VERSION«_

### Copy the following folders and files (from last working version of NetBeans installation):

**File:** _config/Modules/org-netbeans-libs-clank.xml_
**Files:** _config/Modules/org-netbeans-modules-cnd*.xml_
**Files:** _config/Modules/org-netbeans-modules-dlight-kit*.xml_

**Folder**: _config/Preferences/org/netbeans/modules/cnd_
**Folder:** _config/Preferences/org/netbeans/modules/dlight_
**File:** _config/Preferences/org/netbeans/modules/cnd.properties_
 
**File:** _modules/docs/org-netbeans-modules-cnd.jar_
 
**Files:** _modules/locale/org-netbeans-modules-cnd*.jar_
**Files:** _modules/locale/org-netbeans-modules-dlight*.jar_
 
**File:** _modules/ext/org-clank-java.jar_
 
**File:** _modules/org-netbeans-libs-clank.jar_
**Files:** _modules/org-netbeans-modules-cnd*.jar_
**Files:** _modules/org-netbeans-modules-dlight*.jar_
 
**Files:** _update_tracking/org-netbeans-libs-clank.xml_
**Files:** _update_tracking/org-netbeans-modules-cnd*_

Hope it helps 👍

**Comment by vinothv84** [comment]
I will give it a try, Thanks @PetrMiksik

**Comment by vinothv84** [comment]
It didn't work for me. I need remote modules, git, etc. I guess the only way is to wait and hope someone can bring back CND alive. I don't have a necessary skill set, i am just a mere user of Netbeans C/C++.
@bitrunner Just wondering, Is there any remote chance that you might still take up and bring back CND?

**Comment by vinothv84** [comment]
@PetrMiksik Thanks, appreciate the help. But as you can see in your xml files also, remote modules (about 7 or 8 exactly) are disabled. Unfortunately, I need those remote modules, git, etc. :-(

**Comment by PetrMiksik** [comment]
CND lives again :-)
https://github.com/piotrhoppe/netbeans-cnd

**Comment by PetrMiksik** [comment]
CND lives again :-)
https://github.com/piotrhoppe/netbeans-cnd

**Comment by andre-m-coquim-alb** [comment]
Hi @PetrMiksik, what do you mean, will it be available on next release, or doing the procedure you mentioned above, based on that repo, will now make the modules that were not working before work again? (remote modules)

**Comment by PetrMiksik** [comment]
Hi @andre-m-coquim-alb.  You have to install it manually, as described in repo installation procedure. I've tested it with the latest release (NB 31) and it's working :-) Well, it's in early stage of development, so expect some issues, but it looks very good.
The editing features, like code completion, compile errors and go-to-definition are provided by the clangd LSP server, which you have to install separately. The path to clangd server can be set in plugin configuration.

**Comment by andre-m-coquim-alb** [comment]
Thanks, I will be sure to try it and give feedback!
